Classes/Service/AccessService.php

```diff
@@ -31,35 +34,36 @@ public static function getProtectedFileStorages(): array
     }
 
     /**
-     * @param ResourceStorage $storage
-     * @return bool
      * @throws AspectNotFoundException
      */
-    public static function checkStorageAccess(ResourceStorage $storage): bool
+    public static function checkResourceAccess(ResourceStorage $storage, string $identifier): bool
     {
+        // if we have a backend call,
+        // nevermind, as core does access control,
+        // just pass it here
+        if (
+            GeneralUtility::makeInstance(Context::class)
+                ->getPropertyFromAspect('backend.user', 'isLoggedIn')
+        ) {
+            return true;
+        }
         /** @var UserAspect $feUser */
         $feUser = GeneralUtility::makeInstance(Context::class)->getAspect('frontend.user');
+        // early exit, if we don't have a frontend login
         if (!$feUser->isLoggedIn()) {
             return false;
         }
 
-        $storageRecord = $storage->getStorageRecord();
-
-        $groupArray = GeneralUtility::intExplode(',', $storageRecord['fe_group']);
 
-        $datasetGroupArray = [];
-        $db = GeneralUtility::makeInstance(ConnectionPool::class)->getConnectionForTable('fe_users');
-        $fetch = $db
-            ->select(
-                ['*'],
-                'fe_users',
-                ['uid' => $feUser->get('id')]
-            )->fetchAssociative();
-        if ($fetch) {
-            $datasetGroupArray = GeneralUtility::trimExplode(',', $fetch['usergroup']);
-        }
+        // detect current users groups and
+        // folder access groups, if given
+        // otherwise look into file storage
+        $datasetGroupArray = $feUser->getGroupIds();
+        $folderAccess = self::findDirectoryAccess($storage, $identifier);
+        $storageRecord = $storage->getStorageRecord();
+        $groupArray = $folderAccess ? $folderAccess->getFeGroup() : GeneralUtility::intExplode(',', $storageRecord['fe_group']);
 
-        // check if user has group or enabled for every login
+        // check if the user has a related group or enabled for every login
         if (
             in_array(-2, $groupArray) ||
             count(array_diff($groupArray, $feUser->getGroupIds())) < count($groupArray) ||
```

Comment:
micro-optimization:

Do the count for $groupArray into a local variable once before the if and avoid a double count of the same array. It's impossible that it changes between the two condition checks.